Drop module manager and add support for config aggregator

Xerkus opened the pull request:
This PR introduces changes necessary to cut out module manager from MVC application.

Usage originally envisioned for module manager is no longer relevant:
- Package specific autoloader handling is superceded by Composer
- Shared application installations are a thing of the past. Applications are no longer expected to change feature sets at runtime.
- A need for a flexible programmatic application assembly is superseded by dependency injection containers
- Components configuration is immediately available thanks to how composer autoloading works. There is no need to load modules anymore before configuration providers can be gathered and config merged.
- Frontend assets management or package dependencies resolution are no longer needed. Etc. 
-Most used remaining features are a config provider and `onBootstrap()` convenience listener. 

As proven by Mezzio, much simpler laminas-config-aggregator coupled with Container is more than sufficient.  

Changes:
- Module Manager removed along with its dependnecies
- `Laminas\Mvc\Application::init()` is removed in favor of `Laminas\Mvc\ConfigProvider` for laminas-config-aggregator
- `Laminas\Mvc\Service\ConfigFactory` removed as `config` can now be merged and provided to Container before Container is created
- List of extra Application listeners moved from separate application config into main config under the `$config[\Laminas\Mvc\Application::class]['listeners']` key
- Configuration for the various PluginManagers moved from removed `Laminas\ModuleManager\Listener\ServiceListener` into each respective plugin manager factory
- ServiceManagerConfig removed and replaced by ConfigProvider with the goal to eventually remove hard dependency on laminas-servicemanager as a main container.
- ModuleManager's `onBootstrap()` convenience listener is not replaced by any alternative at this time.  

Documentation tracking issue: #58

Xerkus committed:
Introduce ConfigProvider and extract ServiceManagerConfig dependency config

Signed-off-by: Aleksei Khudiakov <aleksey@xerkus.pro>

Xerkus committed:
Extract dependency config from ServiceListenerFactory

Signed-off-by: Aleksei Khudiakov <aleksey@xerkus.pro>

Xerkus committed:
Drop ServiceListenerFactory

Signed-off-by: Aleksei Khudiakov <aleksey@xerkus.pro>

Xerkus committed:
Drop ModuleManagerFactory

Signed-off-by: Aleksei Khudiakov <aleksey@xerkus.pro>

Xerkus committed:
Drop ServiceManagerConfig

Signed-off-by: Aleksei Khudiakov <aleksey@xerkus.pro>

Xerkus committed:
Drop ConfigFactory

Signed-off-by: Aleksei Khudiakov <aleksey@xerkus.pro>

Xerkus committed:
Extract Application::init() listeners bits and drop the rest

Signed-off-by: Aleksei Khudiakov <aleksey@xerkus.pro>

Xerkus committed:
Update and drop tests for removed functionality

Signed-off-by: Aleksei Khudiakov <aleksey@xerkus.pro>

Xerkus committed:
Update integration tests to use ConfigProvider

Signed-off-by: Aleksei Khudiakov <aleksey@xerkus.pro>

Xerkus committed:
Disable bad test in DefaultRenderingStrategyTest

Signed-off-by: Aleksei Khudiakov <aleksey@xerkus.pro>

Xerkus committed:
Remove unused imports

Signed-off-by: Aleksei Khudiakov <aleksey@xerkus.pro>

Xerkus committed:
Drop laminas-modulemanager

Signed-off-by: Aleksei Khudiakov <aleksey@xerkus.pro>

Xerkus commented:
This PR does not concern itself with code quality and intentionally avoids any changes that are not a bare necessity

weierophinney commented on src/Application.php:
```diff
@@ -132,21 +133,13 @@ public function getConfig()
      * @return Application
      */
-    public function bootstrap(array $listeners = [])
+    public function bootstrap()
     {
         $serviceManager = $this->serviceManager;
         $events         = $this->events;
 
-        // Setup default listeners
-        $listeners = array_unique(array_merge($this->defaultListeners, $listeners));
-
-        foreach ($listeners as $listener) {
-            $serviceManager->get($listener)->attach($events);
-        }
```
Why are you removing the ability to pass additional listeners to `bootstrap()`?

To be clear, I'm fine with it (I don't think we document it anywhere), but you'll need to detail it as a feature removal in the changelog as well.

Xerkus replied:
Extra listeners are moved to the application factory since `Application::init()` is going to be dropped.

real change here is that extra listeners to register would migrate from the `application.config.php` to the merged config

weierophinney replied:
:+1:

weierophinney commented on src/Application.php:
```diff
@@ -230,48 +223,6 @@ public function getEventManager()
-     *
-     * If you use this init() method, you cannot specify a service with the
-     * name of 'ApplicationConfig' in your service manager config. This name is
-     * reserved to hold the array from application.config.php.
-     *
-     * The following services can only be overridden from application.config.php:
-     *
-     * - ModuleManager
-     * - SharedEventManager
-     * - EventManager & Laminas\EventManager\EventManagerInterface
-     *
-     * All other services are configured after module loading, thus can be
-     * overridden by modules.
-     *
-     * @param array $configuration
-     * @return Application
-     */
-    public static function init($configuration = [])
-    {
-        // Prepare the service manager
-        $smConfig = isset($configuration['service_manager']) ? $configuration['service_manager'] : [];
-        $smConfig = new Service\ServiceManagerConfig($smConfig);
-
-        $serviceManager = new ServiceManager();
-        $smConfig->configureServiceManager($serviceManager);
-        $serviceManager->setService('ApplicationConfig', $configuration);
-
-        // Load modules
-        $serviceManager->get('ModuleManager')->loadModules();
-
-        // Prepare list of listeners to bootstrap
-        $listenersFromAppConfig     = isset($configuration['listeners']) ? $configuration['listeners'] : [];
-        $config                     = $serviceManager->get('config');
-        $listenersFromConfigService = isset($config['listeners']) ? $config['listeners'] : [];
-
-        $listeners = array_unique(array_merge($listenersFromConfigService, $listenersFromAppConfig));
-
-        return $serviceManager->get('Application')->bootstrap($listeners);
-    }
-
```
Why are you removing this method?

Removal of this MUST be called out in both the CHANGELOG and the documentation, as our own skeleton has consistently used this method in order to initialize and bootstrap the application from aggregated configuration. If it goes away, users need to know what they need to change in their `public/index.php`.

Xerkus replied:
This method no longer provides utility it once did.
With the the module manager dropped, responsibility to setup the application pushed completely out of the scope of this package.

It is now entirely up to skeleton application to provide the default wiring. As initially outlined in the [Zend\Mvc v4 RFC](https://discourse.laminas.dev/t/rfc-zend-mvc-4-design-changes/447) the goal is to move towards current mezzio skeleton application setup. Config+container proved to be flexible enough.

At this time I am working on a new feature that would somewhat approximate laminas-httphanderrunner and would affect how `public/index.php` needs to be changed.

On a side note, I have no immediate intention to provide update for the laminas-modulemanager. I can easily see it retired unless there is a demand for the different set of features not available with the config aggregator.

weierophinney replied:
:+1:

weierophinney commented on src/ConfigProvider.php:
```diff
@@ -0,0 +1,137 @@
+                'EventManager' => false,
+            ],
+            'initializers' => [
+                'EventManagerAwareInitializer' => static function ($first, $second) {
+                    if ($first instanceof ContainerInterface) {
+                        $container = $first;
+                        $instance = $second;
+                    } else {
+                        $container = $second;
+                        $instance = $first;
+                    }
+
+                    if (! $instance instanceof EventManagerAwareInterface) {
+                        return;
+                    }
+
+                    $eventManager = $instance->getEventManager();
+
+                    // If the instance has an EM WITH an SEM composed, do nothing.
+                    if ($eventManager instanceof EventManagerInterface
+                        && $eventManager->getSharedManager() instanceof SharedEventManagerInterface
+                    ) {
+                        return;
+                    }
+
+                    $instance->setEventManager($container->get('EventManager'));
+                },
```
We can simplify this by only allowing laminas-servicemanager v3+ and laminas-eventmanager v3+. Personally, I'd do that as part of this patch to simplify this part of the ConfigProvider.

Alternately, move that code into a dedicated method, and use `Closure::fromCallable()` to provide it as an initializer.

Xerkus replied:
This change would be outside of the scope for the module manager removal.

I do not see a reason to keep EventManager initializer and expect to drop any initializers entirely as a part of the move towards generic PSR containers. I will need to double check container integrations supported by mezzio, I am not sure if it is something supported by all them.

weierophinney replied:
:+1:

weierophinney commented on src/Service/ApplicationFactory.php:
```diff
@@ -27,11 +27,22 @@ class ApplicationFactory implements FactoryInterface
+        $listeners = $container->get('config')[Application::class]['listeners'] ?? [];
+        foreach ($listeners as $listener) {
+            $container->get($listener)->attach($em);
+        }
```
This answers part of my question about the changes to `Application::bootstrap()`. However, you need to document this in the changelog and the manual.

Xerkus replied:
This is meant more of a placeholder at this point that is meant to preserve the functionality. Event system would need to have a more detailed revision and potential utilization of PSR-14 would change this part entirely.

One of the reasons why I opted for skipping documenting the changes in CHANGELOG and providing updated documentation by this PR.

weierophinney replied:
:+1:

weierophinney commented on src/Service/ViewHelperManagerFactory.php:
```diff
@@ -41,39 +40,41 @@ public function __invoke(ContainerInterface $container, $requestedName, array $o
         $urlFactory = $this->createUrlHelperFactory($services);
-        $plugins->setFactory(ViewHelper\Url::class, $urlFactory);
-        $plugins->setFactory('laminasviewhelperurl', $urlFactory);
+        $managerConfig['aliases']['laminasviewhelperurl'] = ViewHelper\Url::class;
```
I think we can drop these "normalized" aliases - users should only be using the FCQN or the class-name-minus-namespace when retrieving helpers.

Xerkus replied:
That change would be outside of the scope of this PR. Too easy to lose track of small details like that during review.

I plan to drop all obsolete aliases in another PR.

weierophinney replied:
:+1:

weierophinney reviewed (commented):
Overall, I think it's a nice step!

I think the one thing missing for me is a write-up in the patch description detailing:

- your proposal for where the config and/or container are built
    - AND what each of those files look like
- what the `public/index.php` looks like

As somebody familiar with Mezzio (clearly!) those details seem obvious, but for somebody not familiar with them, having that information documented here will help them understand how we are evolving the MVC, and what it will look like once complete.

Xerkus committed:
Drop test for empty initial listeners

Application listeners should be registered by the time `Application`
service is retrieved from container. Previously they were registered at
the time of `Application::bootstrap()` call.

Signed-off-by: Aleksei <aleksey@xerkus.pro>

Xerkus commented:
Migration path for existing users would need to be provided via separate package providing necessary bindings between module manager and mvc.
`config()` and `onBootstrap()` are two most often used module methods. First one is covered by config provider while second... I would need to look at event changes first. I do not remember intended details anymore, but it involved PSR event listener providers.

weierophinney commented on src/ConfigProvider.php:
```diff
@@ -0,0 +1,132 @@
+                'configuration' => 'config',
+                'Configuration' => 'config',
+                'HttpDefaultRenderingStrategy' => View\Http\DefaultRenderingStrategy::class,
+                'request' => 'Request',
+                'response' => 'Response',
```
I don't think the request and response should be services anymore in v4, but I'm unsure if that's in-scope for this patch.

Xerkus replied:
They definitely do not  belong to services.
I plan to drop them not only from container but also from bootstrap event. New event would need to be introduced to do request aware bootstap before routing. I think I did that work in zend-mvc but abandoned when we migrated to laminas.

We discussed it long time ago, and it should be on the old rfc: it would make sense to inject them to the run method with laminas-httphandlerrunner when mvc is  migrated to PSR and something similar before that.

weierophinney commented on src/Service/ViewHelperManagerFactory.php:
```diff
@@ -5,13 +5,12 @@
 use Laminas\View\HelperPluginManager;
 
-class ViewHelperManagerFactory extends AbstractPluginManagerFactory
+class ViewHelperManagerFactory implements FactoryInterface
```
Honestly, I'd drop the `implements FactoryInterface` part. Factories only need to be invokable,and it's another explicit binding to laminas-servicemanager instead of a generic PSR-11 implementation.

Xerkus replied:
I agree.

Factories look for extra input parameters that are only supported by service manager. the `$options` specifically, while for the `$name` we do provide decorators for others containers iirc.

I will have a pass dedicated to factories. I think quite a few of them can be safely dropped and replaced by factories provided by components.

weierophinney reviewed (approved):
This looks good.

The only feedback I have is around usage of the laminas-servicemanager `FactoryInterface`; I think we could drop explicitly implementing that interface, and just doing invokables, to help reduce ties to laminas-servicemanager. However, that change is likely out of scope for this PR.

manishranjan-adobe commented:
Hi @froschdesign When do you plan to release version 4.0.*?
